Transfer Verbum app into into jetpack-mu-wpcom

agrullon95 opened the pull request:
## Proposed changes:
Moves [Verbum Comments project](https://github.com/automattic/verbum) into Jetpack mu wpcom plugin. It also has some glue code to handle the transition between the deployed and the upcoming versions.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Does this pull request change what data or activity we track or use?
Not at all. 

## Testing instructions:
* Navigate to `projects/packages/jetpack-mu-wpcom`
* Run `pnpm run build-js` to build Verbum (development) or Run `npm run build-production-js` (production)
* Run `jetpack rsync jetpack mu-wpcom-plugin`  and select your sandbox as the destination (full command example `jetpack rsync mu-wpcom-plugin user@hostname:~/public_html/wp-content/mu-plugins/jetpack-mu-wpcom-plugin/production`
* Make sure you have set `define( 'JETPACK_AUTOLOAD_DEV', true );` in a `mu-plugin/0-sandbox.php`.
* On WPCOM, modify `comment-experience-loader.php` and make `load_verbum_comments` return early.
* The comment experience should continue to function, and it should be Verbum (not Highlander or core's).
* Verify further that the new experience is in charge by checking the path of the `verbum-comment.js` bundle in Network tab.
* Try your best to break the comment experience. Comment as a guest, **comment from Atomic**, and log in using FB.

agrullon95 committed:
Add Verbum Comments in Jetpack-mu-wpcom

agrullon95 committed:
changelog

agrullon95 commented on projects/packages/jetpack-mu-wpcom/webpack.config.js:
```diff
@@ -19,7 +22,28 @@ module.exports = {
+			h: [ 'preact', 'h' ],
+			Fragment: [ 'preact', 'Fragment' ],
+		} ),
+		new CopyPlugin( {
```
Copies Verbum php files into `build/verbum-comments` folder.

agrullon95 commented on projects/packages/jetpack-mu-wpcom/webpack.config.js:
```diff
@@ -34,7 +58,10 @@ module.exports = {
 			// Handle CSS.
-			jetpackConfig.CssRule(),
+			jetpackConfig.CssRule( {
+				extensions: [ 'css', 'sass', 'scss' ],
```
Handles scss conversion to css

agrullon95 commented on projects/packages/jetpack-mu-wpcom/webpack.config.js:
```diff
@@ -19,7 +22,28 @@ module.exports = {
+	plugins: [
+		...jetpackConfig.StandardPlugins( {
+			DependencyExtractionPlugin: { injectPolyfill: true },
+			MiniCssExtractPlugin: { filename: '[name]/[name].css' },
```
Outputs CSS in the same build directory as the entry.

For Verbum, css will output in `build/verbum-comments`.

agrullon95 commented on projects/packages/jetpack-mu-wpcom/babel.config.json:
```diff
@@ -0,0 +1,11 @@
+{
+	"plugins": [
+		[
+			"@babel/plugin-transform-react-jsx",
```
Converts JSX to react function calls.

<div>Example<div> -> React.createElement("div", null, "Example");

agrullon95 committed:
Add dynamic import support to webpack config

agrullon95 committed:
Add logi to load new Jetpack Verbum Comment class if its not already loaded in WPCOM

agrullon95 committed:
Add Verbum Comments project into jetpack-mu-wpcom features directory

agrullon95 committed:
Remove Verbum editor folder since it will be loaded from new package

agrullon95 commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/vite.config.ts:
```diff
@@ -0,0 +1,26 @@
+import preact from '@preact/preset-vite';
```
This file will no longer be needed (using it as a reference) after I'm done converting verbumBundleSize to webpack.config.js:

```
verbumBundleSize: spawnSync( 'gzip', [ '-c', 'dist/index.js' ] ).stdout.toString().trim()
			.length,
```

agrullon95 commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/src/utils.ts:
```diff
@@ -0,0 +1,249 @@
+export function isFastConnection() {
+	// TODO: Need to get bundle size in webpack.
+	// Temporarily setting to true
+	return true;
```
Temporarily setting to true until `verbumBundleSize` is fixed and added to webpack.config.js

agrullon95 replied:
Removed changes to retrieve bundle size and hardcoded to 30000

agrullon95 commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/src/components/editor-placeholder.tsx:
```diff
@@ -0,0 +1,32 @@
+					style={ { padding: '10px 20px' } }
+				>
+					{ loading ? (
+						<CustomLoadingSpinner />
```
Added custom spinner to remove dependency on `@wordpress/components` and reduce project size.

heavyweight replied:
that's great!

agrullon95 commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -211,4 +213,16 @@ public static function load_marketplace_products_updater() {
+	 * Load Verbum Comments.
+	 */
+	public static function load_verbum_comments() {
+		if ( class_exists( 'Verbum_Comments' ) ) {
```
class_exists condition will be removed when this is finalized, merged into WPCOM, removed from mu-plugins/verbum and we're ready to transfer over to Verbum_Comments inside of jetpack-mu-wpcom plugin.

agrullon95 committed:
Add plugin to replace verbumBundleSize global variable value

agrullon95 committed:
Use GetVerbumBundleSizePlugin in webpack config

agrullon95 committed:
Update verbum comments code

agrullon95 committed:
Add custom loading spinner

agrullon95 committed:
remove extra testing custom spinner

heavyweight reviewed (commented):
Double checked the copied files now.
Didn't find any issues.

heavyweight committed:
Update package.json to include required verbum dependencies

heavyweight commented:
I couldn't build this, so I updated package.json

agrullon95 committed:
Merge branch 'trunk' into add/verbum-comment-into-jetpack-mu-wpcom

agrullon95 committed:
update package.json

agrullon95 committed:
Update webpack config

agrullon95 committed:
merge babel config

agrullon95 committed:
Ignore verbum-comments.js build in 18nCheckPlugin

agrullon95 committed:
Move get-verbum-bundle-size-plugin location

agrullon95 committed:
Merge branch 'trunk' into add/verbum-comment-into-jetpack-mu-wpcom

anomiex commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/package-lock.json:
We should not have npm lock files in subdirectories. Integrate the necessary stuff into the package's top-level `package.json` please.

agrullon95 replied:
Thanks. Removing this file and moving necessary dependencies to jetpack-mu-wpcom-plugin package.json.

agrullon95 committed:
Remove i18CheckPlugin filter

alshakero commented on projects/packages/jetpack-mu-wpcom/custom-plugins/get-verbum-bundle-size-plugin.js:
That's not worth it. Let's just hardcode a 30K bundle size.

agrullon95 replied:
Removed and hardcoded value for verbumBundleSize

alshakero commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/composer.json:
doesn't seem to belong here.

agrullon95 replied:
Removed file

alshakero commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/phpcs.xml.dist:
Probably doesn't belong here. Jetpack most probably has one.

agrullon95 replied:
Removed file

alshakero commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/pre-push.sh:
Let's remove this for now. We can't block Jetpack push due to one small app. We'll have to figure a better time to run the tests.

agrullon95 replied:
Removed file

anomiex replied:
It'd probably be best to do tests in CI rather than pre-push. Particularly for something as heavy-weight as E2E tests.

PHPUnit tests are already being run for the package as a whole. There aren't currently any jest tests, but those would be straightforward enough to add; some details [here](https://github.com/Automattic/jetpack/blob/trunk/docs/monorepo.md#javascript-tests).

E2E test setup is documented [here](https://github.com/Automattic/jetpack/tree/trunk/tools/e2e-commons#setup-e2e-tests-for-your-jetpack-plugin), but that would probably want to be set up in `projects/plugins/mu-wpcom-plugin/` rather than the package. Or you could add to the Calyspo tests for Simple and Atomic sites.

alshakero commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/src/i18n.ts:
Since this is in Jetpack, we can probably using proper translations function on the JS level and remove all the strings mapping from the `index.php` file. Not sure, but please check if other apps in wpcom-mu-plugin have access to `@wordpress/i18n`. Although this can be done in a follow up.

agrullon95 replied:
I'll note this down and save for follow up

alshakero commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/.eslintrc:
I don't think we need our own linting rules. Jetpack's should be great.

alshakero commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/.gitignore:
Same. Jetpack probably ignres all of these.

anomiex replied:
Some, but not all.

OTOH, we don't use npm or yarn or lerna so those don't need to be ignored. In general I'd probably delete this and add anything that's actually needed (probably nothing?) to `projects/packages/jetpack-mu-wpcom/.gitignore`.

alshakero commented on projects/packages/jetpack-mu-wpcom/webpack.config.js:
```diff
@@ -23,7 +31,27 @@ module.exports = {
+				},
+			],
+		} ),
+		new GetVerbumBundleSizePlugin( {} ),
```
Let's nuke this.

agrullon95 replied:
Removed this and hardcoded 30000

alshakero reviewed (changes requested):
Left some comments.

agrullon95 committed:
Remove unnecssary files

agrullon95 committed:
Hardcode verbumBundleSize to 30000

agrullon95 committed:
Moving Verbum package.json dependencies to jetpack-mu-wpcom

anomiex committed:
Hack around preact having `__` methods internally

This is the best I can come up with to avoid it making i18n-check-webpack-plugin unhappy.

agrullon95 committed:
index.php lint changes

agrullon95 committed:
Update verbum php usage/import

agrullon95 committed:
Fix lint issues for gutenberg comments php

agrullon95 committed:
update verbum ombed class

agrullon95 committed:
Update verbum auth class lint issues

anomiex committed:
Try something a little less hacky to hack preact

agrullon95 committed:
Fix html_entity_decode lint issue

agrullon95 committed:
php lint fixes

renancarvalho committed:
Fix lint issues part 1

renancarvalho committed:
Fix eslint event issue

alshakero committed:
Turn off `testing-library/prefer-screen-queries`

alshakero commented on projects/packages/jetpack-mu-wpcom/webpack.config.js:
```diff
@@ -65,7 +65,14 @@ module.exports = {
+				includeNodeModules: [ 'preact' ],
+				babelOpts: {
+					configFile: false,
+					plugins: [ [ 'babel-plugin-transform-rename-properties', { rename: { __: '__ǃ' } } ] ],
```
This is beautiful. Thanks @anomiex!

alshakero committed:
Linting

agrullon95 committed:
update util.ts lint

agrullon95 committed:
Add useSocialLogin hook JSDoc and update conditional useEffect

alshakero committed:
Lint ToggleControl

agrullon95 committed:
Update handleBeforeUnload to use preact effect

alshakero committed:
More linting

alshakero committed:
More linting

alshakero committed:
Fix handleBeforeUnload

alshakero committed:
Merge branch 'trunk' into add/verbum-comment-into-jetpack-mu-wpcom

alshakero committed:
Bring over https://github.com/Automattic/verbum/pull/141

alshakero committed:
Merge branch 'trunk' into add/verbum-comment-into-jetpack-mu-wpcom

alshakero reviewed (dismissed):
Looks great to me now!

alshakero commented:
Hi @daledupreez! Would you mind taking `error-reporting` for a smoke test? We changed the webpack config and I'm trying to make sure things are working well.

heavyweight commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/class-verbum-comments.php:
```diff
@@ -0,0 +1,594 @@
+	 * @return void
+	 */
+	public function clear_fb_cookies() {
+		setcookie( 'wpc_fbc', ' ', time() - YEAR_IN_SECONDS, COOKIEPATH, COOKIE_DOMAIN, false, true );
```
Shouldn't this be
```suggestion
		setcookie( 'wpc_fbc', ' ', time() - YEAR_IN_SECONDS, COOKIEPATH, COOKIE_DOMAIN, true, false );
```
<img width="1164" alt="Screenshot 2024-01-22 at 14 03 33" src="https://github.com/Automattic/jetpack/assets/7000684/570afed3-a947-4b2c-b6f1-ec5bef82fe74">

https://www.php.net/manual/en/function.setcookie.php

heavyweight replied:
The equivalent version of what we have on prod would be with `false`, `false` which are the defaults

agrullon95 replied:
I updated the httponly value to true based on the php lint recommendations.
We should probably update both values to true. Are you aware of any issues that could possibly arise from these changes?

heavyweight commented on projects/packages/jetpack-mu-wpcom/src/features/verbum-comments/src/utils.ts:
```diff
@@ -0,0 +1,254 @@
+ */
+export function isFastConnection() {
+	// Hardcoding the size of the bundle.
+	const bytes = 30000;
```
prod build is 232kb for me and 51.6kb are transferred in the networks tab.
how do we plan to manage this value?

agrullon95 replied:
We don't have a solution for calculating and setting this value as of now.
I removed my initial solution since it wouldn't be efficient enough. However, we still need to figure out a way to either calculate this value or another method to determine what a "fast connection" would be.

agrullon95 replied:
>51.6kb are transferred in the networks tab.

I see 51.3kb transferred if I use the dev build and 29.5kb with the production build.
Can you check again?

heavyweight replied:
I'm getting the same now
<img width="887" alt="Screenshot 2024-01-22 at 18 38 14" src="https://github.com/Automattic/jetpack/assets/7000684/85b1e51d-82db-4c34-bfb6-f2b288e369dc">

agrullon95 commented:
Hey @fushar can you please smoke test the `block-theme-previews` feature? I see that you recently added it to jetpack-mu-wpcom. We made a few changes to the webpack config and want to ensure we don't cause issues in your code. Thanks!

heavyweight reviewed (commented):
<img width="1246" alt="Screenshot 2024-01-22 at 18 19 32" src="https://github.com/Automattic/jetpack/assets/7000684/67e35957-6925-4c22-a8d6-4fdec9512958">
When clicking "continue reading" I get this. This looks like a regression

heavyweight commented:
<img width="1163" alt="Screenshot 2024-01-22 at 18 26 32" src="https://github.com/Automattic/jetpack/assets/7000684/b9307d57-d15d-40fb-9ce6-5fe890bf6a09">
also reproducible by writing a comment, logging in with facebook and clicking "Reply"

agrullon95 commented:
@heavyweight 

> <img alt="Screenshot 2024-01-22 at 18 19 32" width="1246" src="https://private-user-images.githubusercontent.com/7000684/298668214-67e35957-6925-4c22-a8d6-4fdec9512958.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5NDYzNDcsIm5iZiI6MTcwNTk0NjA0NywicGF0aCI6Ii83MDAwNjg0LzI5ODY2ODIxNC02N2UzNTk1Ny02OTI1LTRjMjItYThkNi00ZmRlYzk1MTI5NTgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyMiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjJUMTc1NDA3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YWUyNDhkMWRmZmFkMjNlZjljNGZiZTJmYmQ4YTIzNzRjOTljNjhlYjFjNTc2NDI4MjUxZDI5ZmNmZjRlZjZlNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.mzqSF9qkJVFTVz78ZxOsyC4GPT_C-5-h3OtbwBZW2Cc"> When clicking "continue reading" I get this. This looks like a regression

I'm taking a look at these issues

daledupreez commented:
> Hi @daledupreez! Would you mind taking error-reporting for a smoke test? We changed the webpack config and I'm trying to make sure things are working well.

@fullofcaffeine, could you take a look at this? I honestly don't know how to fully test the `error-reporting` logic!

agrullon95 committed:
Update beforeunload logic

agrullon95 committed:
Move handleClose function order

heavyweight committed:
Update event dispose logic

heavyweight committed:
Merge branch 'trunk' into add/verbum-comment-into-jetpack-mu-wpcom

agrullon95 committed:
update package version

agrullon95 committed:
Fix subscription closeModalHandlercommentUrl issue

agrullon95 committed:
Merge branch 'trunk' into add/verbum-comment-into-jetpack-mu-wpcom

agrullon95 commented:
@heavyweight - I pushed a change to fix the logic when you click outside of the Subscription modal

fushar commented:
> Hey @fushar can you please smoke test the block-theme-previews feature? 

Thanks for the ping! I tested on Simple and the feature looks broken now:

<img width="1159" alt="image" src="https://github.com/Automattic/jetpack/assets/1525580/9783048c-bc65-40cb-a216-4ef085153376">

I don't have the bandwidth to investigate this, could you help? All information should be available in this PR, the only PR that introduced the feature:

- https://github.com/Automattic/jetpack/pull/34935

heavyweight commented:
> > Hey @fushar can you please smoke test the block-theme-previews feature?
> 
> Thanks for the ping! I tested on Simple and the feature looks broken now:
> 
> <img alt="image" width="1159" src="https://private-user-images.githubusercontent.com/1525580/298809043-9783048c-bc65-40cb-a216-4ef085153376.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5OTMwNTMsIm5iZiI6MTcwNTk5Mjc1MywicGF0aCI6Ii8xNTI1NTgwLzI5ODgwOTA0My05NzgzMDQ4Yy1iYzY1LTQwY2ItYTIxNi00ZWYwODUxNTMzNzYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDEyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAxMjNUMDY1MjMzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2IzODlhOGUxYTU4MmJlZWUzM2Q1MTA5NDdmNTlmOWE0YTJjOGVkZjBkMWVjMzdlZGNiYzA1YTkyMTdkYWNkNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.84PqKIVakkITj-wa7tnz-iGnCq6dzm4jga9pxrxOr3M">
> I don't have the bandwidth to investigate this, could you help? All information should be available in this PR, the only PR that introduced the feature:
> 
> * [Block Theme Previews: show education modal when previewing theme #34935](https://github.com/Automattic/jetpack/pull/34935)

Thanks for taking the time.
We will take a look at this

heavyweight commented:
> @heavyweight - I pushed a change to fix the logic when you click outside of the Subscription modal

Checked and works great!

alshakero committed:
Dont touch babel and webpack config for other projects

alshakero committed:
Exclude P2 and forums from Verbum

alshakero committed:
Merge branch 'trunk' into add/verbum-comment-into-jetpack-mu-wpcom

heavyweight reviewed (approved):
LGTM

daledupreez commented on projects/packages/jetpack-mu-wpcom/src/class-jetpack-mu-wpcom.php:
```diff
@@ -218,4 +220,42 @@ public static function load_marketplace_products_updater() {
+	 * @return boolean
+	 */
+	private function should_disable_comment_experience( $blog_id ) {
+		require_once WP_CONTENT_DIR . '/lib/wpforteams/functions.php';
```
Thought: this is going to fail on Atomic sites, as this code only exists on WordPress.com Simple. Or did we already verify that it will work?

anomiex replied:
You're right, it did fail on Atomic sites, leading to the revert in #35194. The second attempt at this in #35196 (specifically e89db3566ffcea92c99d357749e73baa1f4b9aa8) both added a check before the require and added a check to only register the action that calls this method on Simple.
